### ISSUE TITLE
Server setup stuck at "starting watchtower"

### ISSUE THREAD

**Issue by zl1775** [Jigsaw-Code/outline-server#22]
Hi,

On my Ubuntu 17.10, when I run the install_server.sh script, I would get stuck at "> Starting Watchtower ......................... OK". Combing through the script, it seems the line "until curl --cacert "${SB_CERTIFICATE_FILE}" -s "${SB_API_URL}/access-keys" >/dev/null; do sleep 1; done" is giving me trouble because I tried to run this line only but the command seems to be stuck due to the get-out-of-loop condition not met. Let me know if anyone knows how to solve this.

**Comment by pwrdbyjames** [comment]
Same issue over here but on Ubuntu 16.04 running on Google Cloud Platform.

**Comment by 7sDream** [comment]
I think this step is try to confirm that other client can connect to server via your **public** IP address.

If you don't provide `SB_PUBLIC_IP` as a env var, it will be the response of `curl https://ipinfo.io/ip`.

So, if you know your public address you can try

```
export SB_PUBLIC_IP="12.34.56.78"
```

and run install script again.

**Comment by zl1775** [comment]
Right. I think it has to do with the fact that I'm sitting behind a nat, do
the randomly generated port number is not open to the public. Although I
did put my server behind DMZ. So I'm not too sure what's going on at the
moment. Thank you anyways!

**Comment by fortuna** [comment]
The installation process waits for the server to be up and running. It does so by querying with
```
curl --cacert "${SB_CERTIFICATE_FILE}" -s "${SB_API_URL}/access-keys"
```

One theory is that the server is running fine, but ${SB_API_URL} is being blocked by the firewall. Maybe we should use localhost in the test instead, but it's unclear to me if it would solve the problem. Needs further investigation

**Comment by zl1775** [comment]
I suspect it has something to do with the ports being blocked too. I'm not sure but I was wondering if it'd be possible to have UPnP baked in the script, or have the user input in the port and open it up manually?

**Comment by zl1775** [comment]
Ok, I think I figured out the problem. It indeed has to do with the port being blocked behind firewall. I manually set SB_API_PORT to something like "34567" in the script, and opened up this specific port on both my router and ufw. Things are working now.

Suggestion: allow users to choose which port to use or bake UPnP in the script.

**Comment by ieaster1** [comment]
I ran into this as well on Ubuntu 16.04 and needed to add a port forwarding rule on the network switch to allow incoming traffic on a specific port.

I set the port with `export SB_API_PORT="<port>"` and then added the rule on my switch to allow TCP/UDP incoming traffic.  Ran the install script and was good to go. Setting `SB_PUBLIC_IP="127.0.0.1"` would have also solved the issue since it would have prevented the need for the network switch to manage the traffic.

I would theorize that there are two possible issues with the current process.
1. Someone's Network doesn't allow for NAT Loopback, or NAT Reflection, than it won't properly route the traffic
2. Most Networks will be configured to block incoming traffic on the port range being randomly generated

**Comment by 4044ever** [comment]
Doh! I made an error with:

`export SB_PUBLIC_IP="12.34.56.78"`

How I can display what I set? Or how can I clear the settings?

**Comment by studentmain** [comment]
- to display: `echo $SB_PUBLIC_IP`
- to clear: `export SB_PUBLIC_IP=""`
- you can use `export SB_PUBLIC_IP="{correct ip}"` directly without clear

4044ever <notifications@github.com> 于 2018年3月27日周二 00:10写道：

**Comment by fortuna** [comment]
The original issue was likely due to a firewall. https://github.com/Jigsaw-Code/outline-server/pull/38 should have clarified what's going on in the installation and provide the information needed for you to unblock the ports.

I'm marking this as closed. Feel free to open a new issue for things that were not resolved with the change.
